### ISSUE TITLE
[R-package] how to plot the  cross validation evaluation result from lgb.cv()?

### ISSUE THREAD

**Issue by Chuang1128** [microsoft/LightGBM#5983]
I want to use the results of lgb.cv, including observation value and predicted value, to draw a scatter diagram, how to do it?

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

I don't quite understand what you're trying to accomplish, but hopefully I can help by explaining how that function works. 

`lgb.cv()` fits multiple models, and returns an object called an `lgb.CVBooster` that contains all of them. That has two attributes that might be useful for your purposes:

* `$boosters` = a list of `lgb.Booster` objects (LightGBM's model object), each of which you can use to create predictions
* `$record_evals` = a list of out-of-fold evaluation metrics, containing the mean (and optionally standard deviation) of evaluation metrics, as of each iteration, summarized over all of those models

The intent of `lgb.cv()` is that you use it to understand the performance of different sets of hyperparameters (captured in `$record_evals`), and then use the best set to re-train on all of your training data. That's why the idea of "the predictions" is ambiguous when discussing the output of `lgb.cv()`.

The example below shows how to obtain predictions from each of the `lgb.Booster` objects in an `lgb.CVBooster`:

```r
library(lightgbm)

data("EuStockMarkets")
trainMat <- data.matrix(
    EuStockMarkets[, c("DAX", "SMI", "CAC")]
)
labelVec <- EuStockMarkets[, "FTSE"]

dtrain <- lightgbm::lgb.Dataset(
    data = trainMat
    , label = labelVec
    , params = list(
        min_data_in_bin = 1L
    )
)
params <- list(
    objective = "regression"
    , metric = "l2"
    , min_data_in_leaf = 1L
    , verbose = 1L
)
cv_booster <- lightgbm::lgb.cv(
    params = params
    , data = dtrain
    , nrounds = 10L
    , nfold = 3L
)

# compare predictions
vizDF <- data.frame(
    actual = labelVec
    , bst1_preds = predict(cv_booster$boosters[[1]]$booster, trainMat)
    , bst2_preds = predict(cv_booster$boosters[[2]]$booster, trainMat)
    , bst3_preds = predict(cv_booster$boosters[[3]]$booster, trainMat)
)
plot(vizDF)
```

<img width="1312" alt="image" src="https://github.com/microsoft/LightGBM/assets/7608904/3744ded6-c5b4-46a0-846c-fec5ef2039c9">

`{lightgbm}` doesn't have any specific data visualization functions which could be described as *"a scatter diagram... [of]... observation value and predicted value"*, but hopefully that example is enough to show you how to get the raw data for the analysis you want to do.

If you have specific questions about the return value of `lgb.cv()`, please ask them and we'd be happy to help.

If you are looking for generic advice on how to make plots in R, please consider posting with the `r` tag at https://stackoverflow.com/questions/tagged/r.

**Comment by Chuang1128** [comment]
thanks!
For example in rf, I type the
Obs <- model$pred$obs (x-axis)
Pred <- model$pred$pred (y-axis)
how to do in LightGBM

**Comment by jameslamb** [comment]
Use `lgb.train()` to produce a model fitted to all your training data, not `lgb.cv()` which produces multiple models on subsets of the data.

See https://lightgbm.readthedocs.io/en/latest/R/reference/predict.lgb.Booster.html or the code I provided above for some examples.

**Comment by Chuang1128** [comment]
But where I can know the CV result and draw the scatter plot?

**Comment by jameslamb** [comment]
Sorry, you will have to be more precise about what you mean by "the CV result".

I explained in https://github.com/microsoft/LightGBM/issues/5983#issuecomment-1637360856 how to interpret the contents of the object returned by `lgb.cv()` and how to produce predictions from each of the individual models inside that object.

**Comment by Chuang1128** [comment]
I am sorry. I'm trying now.

**Comment by Chuang1128** [comment]
Thank you, I solved the problem.
But I have a question, what is the 'cv_booster$boosters[[1]]$booster' mean?

**Comment by jameslamb** [comment]
I did try to answer that in https://github.com/microsoft/LightGBM/issues/5983#issuecomment-1637360856, but I'll rephrase it.

`lgb.cv()` splits the data into training and test sets `nfold` times. For each training and test set combination, it trains a model (called a "booster" in LightGBM).

The `$boosters` attribute of the object returned by that function contains each of those models, which will be slightly different because they were fitted to diferent subsets of the training data.

So, for example, `lgb.cv(nfold=3)` will return an object with 3 fitted models in its `$boosters` attribute.

In that example, `cv_booster$boosters[[1]]$booster` means "the model fit to the first subset of the data".

**Comment by Chuang1128** [comment]
Thank you, let me think a moment.

**Comment by Chuang1128** [comment]
I got it! Thank you.
